corehq/motech/dhis2/models.py

```diff
@@ -252,9 +254,109 @@ def get_dataset(
+                        "period": <period>,
+                        "dataValues": [...]
+                    }
+
+                Since 'completeDate' is specified directly on the `dataset` dictionary, we look there.
+                Several cases should be considered:
+                A) 'orgUnit' and 'period' is static, i.e. already on same level as 'completeDate'.
+                    - No further processing, simply add 'dataValues' to dataset and return dataset as is.
+                B) 'orgUnit' is static and 'period' is dynamic, i.e. 'period' is specified in `datavalues_list`.
+                    - Go through 'datavalues_list' and group by 'period' so that a list of items can be constructed
+                    such that 'period' sits on the same level as 'completeDate'
+                C) 'period' is static and 'orgUnit' is dynamic, i.e. 'period' is specified in `datavalues_list`.
+                    - Same as B), except 'period' is not 'orgUnit'.
+                D) both 'period' and 'orgUnit' is dynamic.
+                    - Same as B and C, except both 'orgUnit' and 'period' should be considered.
+
+            Return:
+                A list of grouped 'datavalues_sets', where each 'set' corresponds to the format
+                specified in "Considerations" above.
+    """
+    if dataset.get('orgUnit') and dataset.get('period'):
+        dataset['dataValues'] = datavalues_list
+        return [dataset]
+
+    group_by_items = []
+    if dataset.get('orgUnit') and not dataset.get('period'):
+        # Need to group datavalues_list by period
+        group_by_items = ['period']
+
+    if dataset.get('period') and not dataset.get('orgUnit'):
+        # Need to group datavalues_list by orgUnit
+        group_by_items = ['orgUnit']
+
+    if not dataset.get('orgUnit') and not dataset.get('period'):
+        # Need to group datavalues_list by orgUnit and period
+        group_by_items = ['orgUnit', 'period']
+
+    return get_grouped_datavalues_sets(
+        group_by=group_by_items,
+        template_dataset=dataset,
+        data_list=datavalues_list
+    )
+
+
+def get_grouped_datavalues_sets(group_by, template_dataset, data_list):
+    """
+        'group_by' will mostly be a list containing just 1 item, but in one instance it contains 2 items, thus
+        all list operations concerning the 'group_by' argument would take a very short time to execute.
+    """
+    def get_grouped_data(data, group_by_keys):
+        data_list_temp = copy.deepcopy(data)
+
+        while len(data_list_temp) > 0:
+            group_by_values_ = [data_list_temp[0][group_by_key] for group_by_key in group_by_keys]
+
+            grouped_data = []
+            iterable_data_list = copy.deepcopy(data_list_temp)
+
+            for data_item in data_list_temp:
+                data_item_key_values = [data_item[group_by_key] for group_by_key in group_by_keys]
+
+                if sorted(data_item_key_values) == sorted(group_by_values_):
```

Comment:
No need to sort these, because keys in `group_by_keys` will always be iterated in the same order. (In fact, if "period" values had the same format as "orgUnit" values, this would break.)